### PR TITLE
Review with focus on extracting demo data

### PR DESCRIPTION
**Pull request by maurolepore**
This PR reviews the MVP code for product-carbon-transition-risk.

### CODE REVIEW

File: inst/extdata/mvp/product-carbon-transition-risk_review.Rmd [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+- a column with production emissions
+- a column indicating the percentile relative to (i) all products with
+  same unit (ii) products in same sector (iii) products in same segment
+- a column indicating whether the product has “low”, “medium” or “high”
+  relative production emissions
+
+The product carbon transition risk indicator consists of 2 main steps:
+
+    - Step 1: Calculating the relative carbon footprint per product
+    - Step 2: Aggregating on the company-level
```
Comment (maurolepore) [inline inst/extdata/mvp/product-carbon-transition-risk_review.Rmd:67]: This suggests we'll need at least two main functions but maybe more. It's better to have many small focused functions that a single big one that does it all. Smaller functions are easier to understand, document, and test. To achieve the complexity the process requires we can simply combine multiple small functions. In pseudo code it might look like this:

```r
emissions_data |>
  calculate_carbon_footprint() |>
  aggregate_carbon_footprint(companies_data)
```

File: inst/extdata/mvp/product-carbon-transition-risk_review.Rmd [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+``` {r}
+library(rlang)
+library(fs)
+library(tidyr)
+library(readr)
+library(dplyr)
+library(graphics)
+library(tiltIndicator)
+
+set.seed(123)
```
Comment (maurolepore) [inline inst/extdata/mvp/product-carbon-transition-risk_review.Rmd:91]: This helps to make the random processes deterministic, i.e. so we always get the same result. But it's best to avoid random processes completely, and create toy data outside the main process.

File: inst/extdata/mvp/product-carbon-transition-risk_review.Rmd [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+
+data_ecoinvent <- params$input1 %||% 
+    read_csv(pctr_path("input_data", "ecoinvent_activities.csv"))
```
Comment (maurolepore) [inline inst/extdata/mvp/product-carbon-transition-risk_review.Rmd:94]: This looks a little weird. It's a small tweak to allow me to develop a function that wraps this .Rmd file, so I can test it and eventually refactor the code in this .Rmd file while ensuring I'm not introducing bugs.

``` r
devtools::load_all()
#> ℹ Loading tiltIndicator
options(readr.show_col_types = FALSE)

input1 <- readr::read_csv(pctr_path("input_data", "ecoinvent_activities.csv"))
input2 <- readr::read_csv(pctr_path("input_data", "MVP_sample_dataset.csv"))

pctr_company_scores(input1, input2)
#> # A tibble: 27 × 3
#>    company_id                                                    risk_…¹ risk_…²
#>    <chr>                                                         <chr>     <dbl>
#>  1 powerstartmanufacturinggmbh_00000005460100827307001           low       0.25 
#>  2 powerstartmanufacturinggmbh_00000005460100827307001           medium    0.5  
#>  3 powerstartmanufacturinggmbh_00000005460100827307001           high      0.25 
#>  4 autohausmanfredmullergmbh_00000004907723001                   low       0.3  
#>  5 autohausmanfredmullergmbh_00000004907723001                   medium    0.5  
#>  6 autohausmanfredmullergmbh_00000004907723001                   high      0.2  
#>  7 ebtor_00000005324899621837001                                 low       0.4  
#>  8 ebtor_00000005324899621837001                                 medium    0.4  
#>  9 ebtor_00000005324899621837001                                 high      0.2  
#> 10 bordelandlandwirtschaftlichegesellschaftmbh_00000004871654001 low       0.182
#> # … with 17 more rows, and abbreviated variable names ¹​risk_groups,
#> #   ²​risk_shares
```

<sup>Created on 2023-03-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: inst/extdata/mvp/product-carbon-transition-risk_review.Rmd [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+names(data_short)[names(data_short) == "Activity Name"] <- "activity"
+names(data_short)[names(data_short) == "Geography"] <- "geo"
+names(data_short)[names(data_short) == "Time Period"] <- "time"
+names(data_short)[names(data_short) == "Special Activity Type"] <- "special_act_type"
+names(data_short)[names(data_short) == "Sector"] <- "sec"
+names(data_short)[names(data_short) == "ISIC Classification"] <- "isic_class"
+names(data_short)[names(data_short) == "ISIC Section"] <- "isic_sec"
+names(data_short)[names(data_short) == "Reference Product Name"] <- "ref_prod"
+names(data_short)[names(data_short) == "CPC Classification"] <- "cpc_class"
+names(data_short)[names(data_short) == "Unit"] <- "unit"
```
Comment (maurolepore) [inline inst/extdata/mvp/product-carbon-transition-risk_review.Rmd:124]: There are simpler ways to solve this but let's not worry about that for now.

File: inst/extdata/mvp/product-carbon-transition-risk_review.Rmd [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+library(fs)
+library(tidyr)
+library(readr)
+library(dplyr)
+library(graphics)
+library(tiltIndicator)
+
+set.seed(123)
+
+data_ecoinvent <- params$input1 %||%
```
Comment (maurolepore) [inline inst/extdata/mvp/product-carbon-transition-risk_review.Rmd:93]: This looks a little weird. It's a small tweak to allow me to develop a function that wraps this .Rmd file, so I can test it and eventually refactor the code in this .Rmd file while ensuring I'm not introducing bugs.

``` r
devtools::load_all()
#> ℹ Loading tiltIndicator
options(readr.show_col_types = FALSE)

input1 <- readr::read_csv(pctr_path("input_data", "ecoinvent_activities.csv"))
input2 <- readr::read_csv(pctr_path("input_data", "MVP_sample_dataset.csv"))

pctr_company_scores(input1, input2)
#> # A tibble: 27 × 3
#>    company_id                                                    risk_…¹ risk_…²
#>    <chr>                                                         <chr>     <dbl>
#>  1 powerstartmanufacturinggmbh_00000005460100827307001           low       0.25 
#>  2 powerstartmanufacturinggmbh_00000005460100827307001           medium    0.5  
#>  3 powerstartmanufacturinggmbh_00000005460100827307001           high      0.25 
#>  4 autohausmanfredmullergmbh_00000004907723001                   low       0.3  
#>  5 autohausmanfredmullergmbh_00000004907723001                   medium    0.5  
#>  6 autohausmanfredmullergmbh_00000004907723001                   high      0.2  
#>  7 ebtor_00000005324899621837001                                 low       0.4  
#>  8 ebtor_00000005324899621837001                                 medium    0.4  
#>  9 ebtor_00000005324899621837001                                 high      0.2  
#> 10 bordelandlandwirtschaftlichegesellschaftmbh_00000004871654001 low       0.182
#> # … with 17 more rows, and abbreviated variable names ¹​risk_groups,
#> #   ²​risk_shares
```

Note I'm no longer using `here::here()` but instead using a new helper `pctr_path()` which looks for data outside this repo (the demo data is still way too large to fit in this package). If you want to run this .Rmd file you'll need to place the data under under `~/Downloads/tilt/tiltIndicator/pctr`.

``` r
devtools::load_all()
#> ℹ Loading tiltIndicator

pctr_path()
#> /home/rstudio/Downloads/tilt/tiltIndicator/pctr
fs::dir_tree(pctr_path())
#> /home/rstudio/Downloads/tilt/tiltIndicator/pctr
#> ├── ep
#> │   ├── categories.csv
#> │   ├── categories_lookup.csv
#> │   ├── companies.csv
#> │   ├── products.csv
#> │   └── products_lookup.csv
#> ├── input_data
#> │   ├── MVP_sample_dataset.csv
#> │   └── ecoinvent_activities.csv
#> └── output_data
#>     ├── company_scores.csv
#>     ├── mvp_pctr_output.csv
#>     └── mvp_product_carbon_tr_output.csv
```

File: inst/extdata/mvp/product-carbon-transition-risk_review.Rmd (hunk 3 above)
Comment (maurolepore) [inline inst/extdata/mvp/product-carbon-transition-risk_review.Rmd:94]: This dataset is way too large. The output has only a few rows, so most of this input data is unused. Work backwards from the output. See which activities you need and pick just those. You may be able to reduce the data from 21K rows to 200 rows or less. Also remove the columns you don't use in the example. You may also clean the name outside this main script. The input data of the main process should already be clean.

File: inst/extdata/mvp/product-carbon-transition-risk_review.Rmd [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+names(data_short)[names(data_short) == "Unit"] <- "unit"
+```
+
+## 3. Add column “production_emissions_per_product”
+
+``` {r}
+# create column with sample emission data, using "rnorm" with normal distribution
+values <- rnorm(nrow(data_short), mean = 6, sd = 1.5)
+data_short <- data_short %>%
+  mutate(production_emissions = values, emissions_unit = "per product unit")
```
Comment (maurolepore) [inline inst/extdata/mvp/product-carbon-transition-risk_review.Rmd:133]: Do this outside this main proces and add the pre-fabricated column already into the demo data. To keep a record of how you did it, you can keep the script in data-raw/name-of-resulting-clean-dataset.R.

File: inst/extdata/mvp/product-carbon-transition-risk_review.Rmd [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+  mutate(score_unit_isic = case_when(
+    perc_unit_isic < 0.3 ~ "low",
+    perc_unit_isic >= 0.3 & perc_unit_isic < 0.7 ~ "medium",
+    perc_unit_isic >= 0.7 ~ "high",
+  ))
+
+# show distributions (each score should have same probability for score_all)
+barplot(prop.table(table(data_short$score_all)))
+barplot(prop.table(table(data_short$score_unit)))
+barplot(prop.table(table(data_short$score_unit_isic)))
```
Comment (maurolepore) [inline inst/extdata/mvp/product-carbon-transition-risk_review.Rmd:187]: This suggests we need a additional function that creates a plot. A function should either compute something (e.g. `company_scores`) or have a side effect (e.g. plot or save a file) but not both.

Here is what the code might eventually look like in ggplot2 (but don't worry about it now because you want to deliver just a .csv file with company scores).

image

File: inst/extdata/mvp/product-carbon-transition-risk_review.Rmd [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+      of space in their name. E.g., there is an activity called "treatment_of_tyre_wear_emissions_passenger_car" instead of            "treatment of tyre wear emissions, passenger car" as in the original dataset.
+
+    - Quick solution for now (I don't want to invest time in replacing underscores with spaces and vice versa: I replace the            products with random example from ecoinvent
+
+    - Problem B with dataset: in the sample MVP dataset, some company_name were associated with multiple company_ids (example:          instead of ID number 00001 for each row, excel filled in 00002, 00003, 00004 in the following rows). I adjusted that            manually, i.e. the issue is resolved in the following dataset.
+
+``` {r}
+# load sample data
+data_companies <- params$input2 %||% 
+  read_csv(pctr_path("input_data", "MVP_sample_dataset.csv"))
```
Comment (maurolepore) [inline inst/extdata/mvp/product-carbon-transition-risk_review.Rmd:211]: Again, this weird code allows me to inject data into `pctr_company_scores()`. Don't worry for now. If you have the data in the right place (`"~/Downloads/tilt/...") it should just work.

File: inst/extdata/mvp/product-carbon-transition-risk_review.Rmd [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+
+# select relevant columns
+data_companies <- data_companies %>%
+  select(all_of(c(
+    "Company_ID", "Company_Name", "Products", "activity", "unit_of_product"
+  )))
+
+column_names <- c("company_id", "company_name", "products", "activity", "unit")
+names(data_companies)[c(1, 2, 3, 4, 5)] <- column_names
```
Comment (maurolepore) [inline inst/extdata/mvp/product-carbon-transition-risk_review.Rmd:220]: Eventually we'll rename columns differently but don't worry now.

File: inst/extdata/mvp/product-carbon-transition-risk_review.Rmd [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+# sub sample
+sub_sample <- data_short %>%
+  distinct(activity, .keep_all = TRUE)
+
+# random selection of 82 activities incl. unit
+sub_sample <- sub_sample[sample(c(1:nrow(sub_sample)), size = 82, replace = TRUE), c("activity", "unit")]
+
+# replace activity list with new sample from above
+data_companies$activity <- sub_sample$activity
+data_companies$unit <- sub_sample$unit
```
Comment (maurolepore) [inline inst/extdata/mvp/product-carbon-transition-risk_review.Rmd:233]: Here ends the process of creating `data_companies` -- which is a crucial input of the rest of the process. But `data_companies` should be created outside this main script and imported as a demo dataset. Work on it in data-raw/data_companies.R.

Note some columns have very weird characters that may be problematic. Maybe best to exclude them from the main demo dataset. We can use weird data to test edge cases.

image

File: tests/testthat/test-pctr_company_scores.R [hunk 11]
```diff
@@ -1,14 +1,32 @@
...
+  input2 <- readr::read_csv(pctr_path("input_data", "MVP_sample_dataset.csv"))
+
+  mvp1 <- mvp_path("product-carbon-transition-risk_review.Rmd")
+  out1 <- pctr_company_scores(input1, input2, mvp = mvp1)
+
+  mvp2 <- mvp_path("product-carbon-transition-risk_review.Rmd")
+  out2 <- pctr_company_scores(input1, input2, mvp = mvp2)
+
+  expect_equal(out1, out2)
+})
```
Comment (maurolepore) [inline tests/testthat/test-pctr_company_scores.R:32]: Don't worry about tests yet. But here are the first minimal tests that I am using to compare `company_scores` resulting from the original-ish MVP with `company_scores` resulting from the reviewed MVP.

**Review by maurolepore — COMMENTED** [review]
@Tilmon,

Here are comments to guide your next iteration on your MVP. While we work together on this, I'll try to move a bit with other MVPs.
